core/src/main/scala/kafka/log/Log.scala

```diff
@@ -475,28 +474,34 @@ class Log(@volatile var dir: File,
           maxOffsetInMessages = appendInfo.lastOffset)
 
 
-        // verify that the epoch and sequence are correct before appending
-        pidMap.checkSeqAndEpoch(appendInfo.pid, appendInfo.firstSequence, appendInfo.epoch)
-
-        // now append to the log
-        segment.append(firstOffset = appendInfo.firstOffset,
-          largestOffset = appendInfo.lastOffset,
-          largestTimestamp = appendInfo.maxTimestamp,
-          shallowOffsetOfMaxTimestamp = appendInfo.offsetOfMaxTimestamp,
-          records = validRecords)
+        try {
+          // verify that the epoch and sequence are correct before appending
+          pidMap.checkSeqAndEpoch(appendInfo.pid, appendInfo.firstSequence, appendInfo.epoch)
+          // now append to the log
+          segment.append(firstOffset = appendInfo.firstOffset,
+            largestOffset = appendInfo.lastOffset,
+            largestTimestamp = appendInfo.maxTimestamp,
+            shallowOffsetOfMaxTimestamp = appendInfo.offsetOfMaxTimestamp,
+            records = validRecords)
 
-        // update the PID sequence mapping
-        pidMap.update(appendInfo.pid, appendInfo.lastSequence, appendInfo.epoch, appendInfo.lastOffset)
+          // update the PID sequence mapping
+          pidMap.update(appendInfo.pid, appendInfo.lastSequence, appendInfo.epoch, appendInfo.lastOffset)
 
-        // increment the log end offset
-        updateLogEndOffset(appendInfo.lastOffset + 1)
+          // increment the log end offset
+          updateLogEndOffset(appendInfo.lastOffset + 1)
 
-        trace("Appended message set to log %s with first offset: %d, next offset: %d, and messages: %s"
-          .format(this.name, appendInfo.firstOffset, nextOffsetMetadata.messageOffset, validRecords))
+          trace("Appended message set to log %s with first offset: %d, next offset: %d, and messages: %s"
+            .format(this.name, appendInfo.firstOffset, nextOffsetMetadata.messageOffset, validRecords))
 
-        if (unflushedMessages >= config.flushInterval)
-          flush()
+          if (unflushedMessages >= config.flushInterval)
+            flush()
 
+        } catch {
+          case e: DuplicateSequenceNumberException =>
```

Comment:
Do we need to propagate this error after catching? I'm not sure I see how the error code gets set in the response. Maybe we need to add an `Errors` field to `LogAppendInfo`?